File: ios/Epyc/Core/Repositories/DrawingRepository.swift
import Foundation
import FirebaseFirestore
import EpycCore

/// Owns Firestore I/O for drawings. Events live in a subcollection ordered by a
/// monotonic sequence number, mirroring the legacy append-by-index list.
struct DrawingRepository {

    /// Creates an empty drawing document and returns its id.
    func createDrawing(authorUid: String) async throws -> String {
        let ref = FS.drawings.document()
        try await ref.setData([
            "createdAtMs": Date().timeIntervalSince1970 * 1000,
            "authorUid": authorUid
        ])
        return ref.documentID
    }

    /// Appends an event at a specific sequence index (doc id = zero-agnostic seq).
    func appendEvent(drawingId: String, seq: Int, event: DrawingEvent) async throws {
        var data: [String: Any] = ["type": event.type.rawValue, "ts": event.ts]
        if let p = event.pathName { data["pathName"] = p }
        if let x = event.x { data["x"] = x }
        if let y = event.y { data["y"] = y }
        try await FS.drawingEvents(drawingId).document(String(format: "%08d", seq)).setData(data)
    }

    /// Loads all events once (used to seed a recording canvas re-entry).
    func loadEvents(drawingId: String) async throws -> [DrawingEvent] {
        let snapshot = try await FS.drawingEvents(drawingId).order(by: FieldPath.documentID()).getDocuments()
        return snapshot.documents.compactMap { Self.decode($0.data()) }
    }

    /// Streams events in order for live replay.
    func observeEvents(drawingId: String) -> AsyncStream<[DrawingEvent]> {
        snapshotStream { handler in
            FS.drawingEvents(drawingId)
                .order(by: FieldPath.documentID())
                .addSnapshotListener { snapshot, error in
                    if let error { handler(.failure(error)); return }
                    guard let snapshot else { return }
                    handler(.success(snapshot.documents.compactMap { Self.decode($0.data()) }))
                }
        }
    }

    private static func decode(_ data: [String: Any]) -> DrawingEvent? {
        guard let typeStr = data["type"] as? String,
              let type = DrawingEventType(rawValue: typeStr),
              let ts = data["ts"] as? Double else { return nil }
        return DrawingEvent(type: type,
                            ts: ts,
                            pathName: data["pathName"] as? String,
                            x: data["x"] as? Double,
                            y: data["y"] as? Double)
    }
}

File: flutter/lib/src/services/firestore/drawing_repository.dart
import 'package:cloud_firestore/cloud_firestore.dart';

import '../../models/drawing_event.dart';

/// Firestore-backed port of `src/providers/drawing-model/drawing-model.ts`.
///
/// Schema: `drawings/{drawingId}` with an ordered `events` subcollection. Each
/// event carries a monotonic `seq` for replay ordering (replaces the RTDB
/// numeric-index list).
class DrawingRepository {
  DrawingRepository(this._db);

  final FirebaseFirestore _db;

  CollectionReference<Map<String, dynamic>> get _drawings =>
      _db.collection('drawings');

  CollectionReference<Map<String, dynamic>> _events(String drawingId) =>
      _drawings.doc(drawingId).collection('events');

  /// Port of `createInstance`: creates an empty drawing and returns its id.
  Future<String> createInstance() async {
    final ref = _drawings.doc();
    await ref.set({'createdMs': DateTime.now().millisecondsSinceEpoch});
    return ref.id;
  }

  /// Appends a drawing event. Port of `DrawingEventList.storeDrawingEvent`.
  /// The [seq] preserves recording order on read-back.
  Future<void> appendEvent(String drawingId, DrawingEvent event, int seq) {
    final map = event.toMap()..['seq'] = seq;
    return _events(drawingId).add(map);
  }

  /// Streams all events for a drawing in recording order. Port of
  /// `loadDrawingEvents`, used by the replaying canvas.
  Stream<List<DrawingEvent>> watchEvents(String drawingId) => _events(drawingId)
      .orderBy('seq')
      .snapshots()
      .map((q) => q.docs.map((d) => DrawingEvent.fromMap(d.data())).toList());

  /// One-shot read of all events (for results playback that does not need a
  /// live subscription).
  Future<List<DrawingEvent>> fetchEvents(String drawingId) async {
    final q = await _events(drawingId).orderBy('seq').get();
    return q.docs.map((d) => DrawingEvent.fromMap(d.data())).toList();
  }
}
